tests/foundry/common/abstract/type-data.mjs.test-d.ts

```diff
@@ -0,0 +1,130 @@
+import { expectTypeOf } from "vitest";
+import type { fields } from "../../../../src/foundry/common/data/module.d.mts";
+import type BaseJournalEntryPage from "../../../../src/foundry/common/documents/journal-entry-page.d.mts";
+import type { TypeDataModel } from "../../../../src/foundry/common/abstract/type-data.d.mts";
+
+/* attempting to use the example as a test */
+
+interface QuestSchema extends BaseJournalEntryPage.Schema {
+  description: fields.HTMLField<{ required: false; blank: true; initial: "" }>;
+  steps: fields.ArrayField<fields.StringField<{ required: true }>>;
+}
+
+type BaseQuestData = {
+  // Overrides the schema
+  description: HTMLElement;
+  questName: string;
+};
+type DerivedQuestData = { totalSteps: number };
+
+// Test With specified Base and DerivedData.
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+class QuestModel extends foundry.abstract.TypeDataModel<
+  QuestSchema,
+  BaseJournalEntryPage,
+  BaseQuestData,
+  DerivedQuestData
+> {
+  otherMethod() {}
+
+  prepareBaseData(this: TypeDataModel.PrepareBaseDataThis<this>): void {
+    this.otherMethod();
```

Comment:
Being able to use the full capabilities in `this` was one of the largest points of these adjustments.